libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp

```diff
@@ -0,0 +1,615 @@
+                local_target != Base::UNASSIGNED) {
+              katana::atomicAdd(
+                  c_info[local_target].degree_wt, n_data_degree_wt);
+              katana::atomicAdd(c_info[local_target].size, (uint64_t)1);
+              katana::atomicSub(
+                  c_info[n_data_curr_comm_id].degree_wt, n_data_degree_wt);
+              katana::atomicSub(c_info[n_data_curr_comm_id].size, (uint64_t)1);
+
+              /* Set the new cluster id */
+              n_data_curr_comm_id = local_target;
+            }
+          },
+          katana::loopname("louvain algo: Phase 1"));
+
+      /* Calculate the overall modularity */
+      double e_xx = 0;
+      double a2_x = 0;
+
+      curr_mod = Base::template CalModularity<EdgeWeightType>(
+          graph, c_info, e_xx, a2_x, constant_for_second_term);
+
+      if ((curr_mod - prev_mod) < threshold_mod) {
+        prev_mod = curr_mod;
+        break;
+      }
+
+      prev_mod = curr_mod;
+
+    }  // End while
+    TimerClusteringWhile.stop();
+
+    iter = num_iter;
+
+    c_info.destroy();
+    c_info.deallocate();
+
+    c_update.destroy();
+    c_update.deallocate();
+
+    TimerClusteringTotal.stop();
+    return prev_mod;
+  }
+
+public:
+  katana::Result<void> LouvainClustering(
+      katana::PropertyGraph* pfg, const std::string& edge_weight_property_name,
+      LargeArray& clusters_orig, LouvainClusteringPlan plan) {
+    /*
+     * Construct temp property graph. This graph gets coarsened as the
+     * computation proceeds.
+     */
+    auto pfg_mutable = std::make_unique<katana::PropertyGraph>();
+    katana::LargeArray<uint64_t> out_indices_next;
+    katana::LargeArray<uint32_t> out_dests_next;
+
+    out_indices_next.allocateInterleaved(pfg->topology().num_nodes());
+    out_dests_next.allocateInterleaved(pfg->topology().num_edges());
+
+    auto numeric_array_out_indices =
+        std::make_shared<arrow::NumericArray<arrow::UInt64Type>>(
+            static_cast<int64_t>(pfg->topology().num_nodes()),
+            arrow::MutableBuffer::Wrap(
+                out_indices_next.data(), pfg->topology().num_nodes()));
+    auto numeric_array_out_dests =
+        std::make_shared<arrow::NumericArray<arrow::UInt32Type>>(
+            static_cast<int64_t>(pfg->topology().num_edges()),
+            arrow::MutableBuffer::Wrap(
+                out_dests_next.data(), pfg->topology().num_edges()));
+
+    if (auto r = pfg_mutable->SetTopology(katana::GraphTopology{
+            .out_indices = std::move(numeric_array_out_indices),
+            .out_dests = std::move(numeric_array_out_dests),
+        });
+        !r) {
+      return r.error();
+    }
+
+    auto graph_result = Graph::Make(pfg);
+    if (!graph_result) {
+      return graph_result.error();
+    }
+    Graph graph_curr = graph_result.value();
+
+    /*
+    * Vertex following optimization
+    */
+    if (plan.is_enable_vf()) {
+      [[maybe_unused]] uint64_t num_nodes_to_fix = Base::VertexFollowing(
```

Comment:
It looks like this might be called for the side effect. If that is the case, I would prefer you remove the variable it's assigned to entirely so no one thinks the variable is the important part and deletes the call. A call without an assignment is much more clearly for side-effects.

Reply:
Good point.